domain/port/service/service.go

```diff
@@ -312,3 +237,182 @@ func verifyNoPortRangeConflicts(rangesA, rangesB []network.PortRange) error {
 	}
 	return errors.Errorf("%w: %s", port.ErrPortRangeConflict, conflicts)
 }
+
+// verifyNoColocatedPortRangeConflicts verifies the provided port ranges do not
+// conflict with the port ranges opened on units co-located with the given unit.
+func (s *Service) verifyNoColocatedPortRangeConflicts(
+	ctx domain.AtomicContext, unitUUID string, portRanges []network.PortRange,
+) error {
+	colocatedOpened, err := s.st.GetColocatedOpenedPorts(ctx, unitUUID)
+	if err != nil {
+		return errors.Errorf("failed to get opened ports co-located with unit %s: %w", unitUUID, err)
+	}
+	return verifyNoPortRangeConflicts(portRanges, colocatedOpened)
+}
+
+// ensureEndpoints ensures the given endpoints required to open and close a given
+// set of ports are present for the given unit. Returns all endpoints for the unit
+// after ensuring after adding any.
+func (s *Service) ensureEndpoints(
+	ctx domain.AtomicContext, unitUUID string, openPorts, closePorts network.GroupedPortRanges,
+) ([]port.Endpoint, error) {
+	currentEndpoints, err := s.st.GetEndpoints(ctx, unitUUID)
+	if err != nil {
+		return nil, errors.Errorf("failed to get unit endpoints: %w", err)
+	}
+
+	newEndpointNamesSet := set.NewStrings()
+	for endpoint := range openPorts {
+		newEndpointNamesSet.Add(endpoint)
+	}
+	for endpoint := range closePorts {
+		newEndpointNamesSet.Add(endpoint)
+	}
+	for _, endpoint := range currentEndpoints {
+		newEndpointNamesSet.Remove(endpoint.Endpoint)
+	}
+	newEndpointNames := newEndpointNamesSet.SortedValues()
+
+	var newEndpoints []port.Endpoint
+	if len(newEndpointNames) > 0 {
+		var err error
+		newEndpoints, err = s.st.AddEndpoints(ctx, unitUUID, newEndpointNames)
+		if err != nil {
+			return nil, errors.Errorf("failed to add endpoints for unit %s: %w", unitUUID, err)
+		}
+	}
+	return append(currentEndpoints, newEndpoints...), nil
+}
+
+func (s *Service) reconcileWildcard(
+	ctx domain.AtomicContext, unitUUID string, endpoints []port.Endpoint, wildcardOpened []port.PortRangeUUID, openPorts, closePorts network.GroupedPortRanges,
+) (network.GroupedPortRanges, network.GroupedPortRanges, error) {
+	wildcardOpen, _ := openPorts[WildcardEndpoint]
+	wildcardClose, _ := closePorts[WildcardEndpoint]
+
+	wildcardOpenedSet := map[network.PortRange]bool{}
+	for _, portRange := range wildcardOpened {
+		wildcardOpenedSet[portRange.PortRange] = true
+	}
+
+	// Remove openPorts ranges that are already open on the wildcard endpoint.
+	for endpoint, endpointOpenPorts := range openPorts {
+		if endpoint == WildcardEndpoint {
+			continue
+		}
+		for i, portRange := range endpointOpenPorts {
+			if _, ok := wildcardOpenedSet[portRange]; ok {
+				openPorts[endpoint] = append(openPorts[endpoint][:i], openPorts[endpoint][i+1:]...)
```

Comment:
I'm more of a fan of creating a new type of things you want to keep, rather than modify out what you don't want to keep, esp. because you don't own that original type.